File: plugins/logistics/backend/dto/route_stop_results.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class RouteStopResultUpsertRequest(BaseModel):
    status: str = Field(min_length=1, max_length=20)
    completion_percent: float = Field(ge=0, le=100)
    outcome_type: str = Field(min_length=1, max_length=40)
    driver_note: str | None = Field(default=None, max_length=1000)


class RouteStopResultRead(BaseModel):
    id: str
    session_id: str
    route_stop_id: str
    status: str
    completion_percent: float
    outcome_type: str
    driver_note: str | None = None
    created_at: datetime
    updated_at: datetime

File: plugins/logistics/frontend/api/route-stop-results.ts
import { API_PREFIX } from "./_shared";
import { apiRequest } from "../../../../apps/web/src/shared/api/client";

export type RouteStopResult = {
  id: string;
  session_id: string;
  route_stop_id: string;
  status: string;
  completion_percent: number;
  outcome_type: string;
  driver_note: string | null;
  created_at: string;
  updated_at: string;
};

export type RouteStopResultUpsertPayload = {
  status: string;
  completion_percent: number;
  outcome_type: string;
  driver_note?: string | null;
};

export function listRouteStopResults(sessionId: string) {
  return apiRequest<RouteStopResult[]>(`${API_PREFIX}/vehicle-sessions/${sessionId}/route-stop-results`);
}

export function upsertRouteStopResult(
  sessionId: string,
  routeStopId: string,
  payload: RouteStopResultUpsertPayload
) {
  return apiRequest<RouteStopResult>(
    `${API_PREFIX}/vehicle-sessions/${sessionId}/route-stop-results/${routeStopId}`,
    {
      method: "PUT",
      body: JSON.stringify(payload),
    }
  );
}
